[SPARK-7481] [build] Add spark-hadoop-cloud module to pull in object store access.

steveloughran opened the pull request:
## What changes were proposed in this pull request?

Add a new `spark-hadoop-cloud` module and maven profile to pull in object store support from `hadoop-openstack`, `hadoop-aws` and `hadoop-azure` (Hadoop 2.7+) JARs, along with their dependencies, fixing up the dependencies so that everything works, in particular Jackson.

It restores `s3n://` access to S3, adds its `s3a://` replacement, OpenStack `swift://` and azure `wasb://`.

There's a documentation page, `cloud_integration.md`, which covers the basic details of using Spark with object stores, referring the reader to the supplier's own documentation, with specific warnings on security and the possible mismatch between a store's behavior and that of a filesystem. In particular, users are advised be very cautious when trying to use an object store as the destination of data, and to consult the documentation of the storage supplier and the connector. 

(this is the successor to #12004; I can't re-open it)

## How was this patch tested?

Downstream tests exist in [https://github.com/steveloughran/spark-cloud-examples/tree/master/cloud-examples](https://github.com/steveloughran/spark-cloud-examples/tree/master/cloud-examples)

Those verify that the dependencies are sufficient to allow downstream applications to work with s3a, azure wasb and swift storage connectors, and perform basic IO & dataframe operations thereon. All seems well.


Manually clean build & verify that assembly contains the relevant aws-* hadoop-* artifacts on Hadoop 2.6; azure on a hadoop-2.7 profile.

SBT build: `build/sbt -Phadoop-cloud -Phadoop-2.7 package`
maven build `mvn install -Phadoop-cloud -Phadoop-2.7`


This PR *does not* update `dev/deps/spark-deps-hadoop-2.7` or `dev/deps/spark-deps-hadoop-2.6`, because unless the hadoop-cloud profile is enabled, no extra JARs show up in the dependency list. The dependency check in Jenkins isn't setting the property, so the new JARs aren't visible.

steveloughran committed:
[SPARK-7481] stripped down packaging only module

steveloughran committed:
[SPARK-7481] basic instantiation tests verify that dependency hadoop-azure, hadoop-aws, hadoop-openstack and implicitly their transitive dependencies are resolved. They don't verify all dependency setup, specifically that Jackson versions are consistent; that needs integration testing.

steveloughran committed:
[SPARK-7481] tests restricted to instantiation; logging modified appropriately

steveloughran committed:
[SPARK-7481] declare httpcomponents:httpclient explicitly, as downstream tests which pulled in spark-cloud but not spark-hive were ending up with inconsistent versions. Add a test for the missing class being there too.

steveloughran committed:
[SPARK-7481] update docs by culling section on cloud integration tests; link to remaning docs from top level.

steveloughran committed:
[SPARK-7481]  updated documentation as per review

steveloughran committed:
[SPARK-7481]  SBT will build this now, optionally

steveloughran committed:
[SPARK-7481] cloud POM includes jackson-dataformat-cbor, so that the CP is set up consistently for the later versions of the AWS SDK

steveloughran committed:
[SPARK-7481]  rebase with master; Pom had got out of sync

steveloughran committed:
[SPARK-7481] rename spark-cloud module to spark-hadoo-cloud, in POMs and docs

steveloughran committed:
[SPARK-7841] bump up cloud pom to 2.2.0-SNAPSHOT; other minor pom cleanup

steveloughran committed:
[SPARK-7481] builds against Hadoop shaded 3.x clients failing as direct references to AWS classes failing. Cut them and rely on transitive load through FS class instantation to force the load. All that happens is that failures to link will be slightly less easy to debug.

steveloughran committed:
[SPARK-7481] update 2.7 dependencies to include azure, aws and openstack JARs, transitive dependencies on aws and azure SDKs

steveloughran committed:
[SPARK-7481] add joda time as the dependency. Tested against hadoop branch-2, s3 ireland

steveloughran committed:
SPARK-7481 purge all tests from the cloud module

steveloughran committed:
SPARK-7481 add cloud module to sbt sequence

Change-Id: I3dea2544f089615493163f0fae482992873f9c35

steveloughran committed:
SPARK-7481 break line of mvn XML declaration

Change-Id: Ibd6d40df2bc8a2edf19a058c458bea233ba414fd

steveloughran committed:
SPARK-7481 cloud pom is still JAR (not pom). works against Hadoop 2.6 as well as 2.7, keeping azure the 2.7.x dependency. All dependencies are scoped @ hadoop.scope

Change-Id: I80bd95fd48e21cf2eb4d94907ac99081cd3bd375

steveloughran committed:
SPARK-7481 move to Spark 2.3.0-SNAPSHOT

Change-Id: I91f764aeed7d832df1538453d869a7fd83964d65

steveloughran committed:
tweaked pom; updated docs

Change-Id: I12ea6ed72ffa9edee964c90c862ff4c45bc4f47f

steveloughran committed:
SPARK-7481 strip down the docs to a bare minimum: FS differences, security, spark-specific options + links elsewhere

Change-Id: I7e9efe20d116802a403af875b241b91178078d78

steveloughran committed:
SPARK-7481 doc review

Change-Id: I1923a4b6a959d86aa2c5b3d71faaaf2541d3ba85

steveloughran committed:
review comments

Change-Id: I6a0b0b9f06a4adcdf55ef75161dc1039961bc7a1

steveloughran committed:
SPARK-7481 more proofreading

Change-Id: Ic4804667af8e52b7be11fb00621ad8b69a1d2569

steveloughran committed:
SPARK-7481 proofreading docs

Change-Id: I2b75a2722f0082b916b9be20bd23a0bdc2d36615

srowen commented:
OK, but does this address the comments from the other PR? it didn't look like it (yet) from a quick look.

SparkQA commented:
**[Test build #76390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76390/testReport)** for PR 17834 at commit [`72a03ed`](https://github.com/apache/spark/commit/72a03ed58331813b0ad4bc9517fcc1f23a5eda6f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76390/
Test PASSed.

steveloughran commented:
The last one was on all the doc comments, and believe I've addressed them both with the little typos and by focusing the docs on the main points for Spark users: how stores differ from filesystems, and what it means

The big issue for spark users is "the commit problem", where I've listed the object store behaviours and said "this means things may not work —consult the docs". I'm not being explicit as what works/doesn't work as that's the moving target. 

Right now, I don't trust commits with S3a using V1 or V2 FileOutputCommitter algorithms to work 100% of the time, because they rely on a list consistency which Amazon S3 doesn't guarantee. I could make that a lot clearer, something like

*You cannot reliably use the FileOutputCommitter to commit work to Amazon S3 or Openstack Swift unless there is some form of consistency layer on top*.

That probably is the core concept people need to know: it's not safe without something (EMR, S3Guard, Databricks Commit Service) to give you that consistent view.

Then add pointers to the talks by myself and [Eric Liang](https://www.slideshare.net/databricks/robust-and-scalable-etl-over-cloud-storage-with-apache-spark) on the topic

srowen commented on pom.xml:
```diff
@@ -1145,6 +1150,70 @@
+        the AWS module pulls in jackson; its transitive dependencies can create
+        intra-jackson-module version problems.
+        -->
+      <dependency>
```
Continuing https://github.com/apache/spark/pull/12004#discussion_r113700119 - I think it's useful to put things in the parent when they're used across multiple modules. For something only referenced in one child POM, it is probably less overhead to just put it the dependency in the child rather than split and repeat part of it.

steveloughran replied:
OK, I'll do that. It's easily enough revisited in future.

The one thing that will probably need revisiting jackson dependency, as that's got the highest risk that someone will add it to another module. Though as it is driven by the same jackson.version variable as everywhere else, a duplicate declaration elsewhere isn't going to cause any harm

srowen commented on cloud/pom.xml:
```diff
@@ -0,0 +1,106 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Oh, just noticed this -- should this perhaps be in a directory `hadoop-cloud` to match the profile and module name?

steveloughran replied:
OK

srowen commented on docs/cloud-integration.md:
```diff
@@ -0,0 +1,190 @@
+
+This uses the "version 2" algorithm for committing files, which does less
+renaming than the "version 1" algorithm, though as it still uses `rename()`
+to commit files, it may be unsafe to use.
```
It's hard to give advice here, as we discussed. It's possible to link to external documentation. Here, however, it sounds like you're saying version 2 is recommended, then immediately saying it could be unsafe to use. What does the reader make of that? what's the upside, what's the nature and likelihood of failure?

If there's no clear answer, is this really a reliable recommendation?
Should this perhaps call out the issue and link to a larger discussion instead?

steveloughran replied:
I'll try to think of a better phrasing, saying "if your object store is consistent enough use v2 for speed".

srowen reviewed (changes requested):
OK still a few comments but yes you're right most were addressed

steveloughran committed:
SPARK-7481 module rename, POM movement, docs

*  module is renamed hadoop-cloud in POMs, sbt, docs
*  hadoop-aws/azure/openstack declarations pushed down to hadoop-cloud pom, along with jackson-cbor
* docs around the commit algorithm option make clear that you should only worry about v1 vs v2 if the blobstore is consistent

Change-Id: Ia114bc8cd2ef731d54a83774d9dc2cf9e4c6e7d4

SparkQA commented:
**[Test build #76426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76426/testReport)** for PR 17834 at commit [`b788494`](https://github.com/apache/spark/commit/b788494cb63c91814309cdf22b55d3301292ac66).

steveloughran commented:
I've just pushed up an update which changes the module name; tested in maven and SBT; hadoop cloud JAR dependencies pulled down.

A JAR is created, it's just a stub one. As a result, when you build the assembly with -Phadoop-cloud, you get something from spark mixed in with the hadoop modules
```
assembly/target/scala-2.11/jars/hadoop-aws-2.7.3.jar
assembly/target/scala-2.11/jars/hadoop-azure-2.7.3.jar
assembly/target/scala-2.11/jars/hadoop-client-2.7.3.jar
assembly/target/scala-2.11/jars/hadoop-cloud_2.11-2.3.0-SNAPSHOT.jar
assembly/target/scala-2.11/jars/hadoop-common-2.7.3.jar
assembly/target/scala-2.11/jars/hadoop-hdfs-2.7.3.jar
assembly/target/scala-2.11/jars/hadoop-mapreduce-client-app-2.7
```

Like I've said before, that's a bit dangerous, especially with a hadoop-cloud-projects POM module now upstream. We won't ever produce a hadoop-cloud-projects JAR, so there won't be direct conflict, more potential confusion if people see a jar beginning hadoop-* with different version info.

SparkQA commented:
**[Test build #76426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76426/testReport)** for PR 17834 at commit [`b788494`](https://github.com/apache/spark/commit/b788494cb63c91814309cdf22b55d3301292ac66).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76426/
Test PASSed.

srowen commented:
That's because you also changed the artifact name to `hadoop-cloud-...`. That should remain `spark-hadoop-cloud-...`. It's the directory, module name, and profile ID that should be `hadoop-cloud`. It is at least consistent with how other modules and profiles are named.

steveloughran commented:
OK, now I understand. let me revert that bit of the patch

steveloughran committed:
SPARK-7481: artifact is spark-hadoop-cloud while project is hadoop-cloud. Added an extra callout to the docs "without some form of consistency layer, Amazon S3 cannot
be safely used as the direct destination of work with the normal rename-based committer"

That holds for all Spark-in-EC deployments; EMR has a consistency option, as do/will others; different committers can work reliably.

Change-Id: Ibbf8b1b9de10b5bb83b647cc663ceb970f70ff2d
Testing: SBT and maven, with & without the hadoop-2.7 option.

SparkQA commented:
**[Test build #76457 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76457/testReport)** for PR 17834 at commit [`e173e3f`](https://github.com/apache/spark/commit/e173e3f2a60a8ecc9875dbda24beba793d86d019).

SparkQA commented:
**[Test build #76457 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76457/testReport)** for PR 17834 at commit [`e173e3f`](https://github.com/apache/spark/commit/e173e3f2a60a8ecc9875dbda24beba793d86d019).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76457/
Test PASSed.

srowen commented on docs/cloud-integration.md:
```diff
@@ -0,0 +1,203 @@
+* Rename operations may be very slow and, on failure, leave the store in an unknown state.
+* Seeking within a file may require new HTTP calls, hurting performance. 
+
+How does affect Spark? 
```
Just noticed a small typo: How does this affect Spark?

steveloughran replied:
fixed, "how does this affect Spark?"

srowen commented on docs/cloud-integration.md:
```diff
@@ -0,0 +1,203 @@
+They cannot be used as a direct replacement for a cluster filesystem such as HDFS
+*except where this is explicitly stated*.
+
+Key differences are
```
Nit: I'd end the line with a colon to make it clear it's not dangling

srowen commented on docs/cloud-integration.md:
```diff
@@ -0,0 +1,203 @@
+### Important: Cloud Object Stores are Not Real Filesystems
+
+While the stores appear to be filesystems, underneath
+they are still object stores, [and the difference is significant](http://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/filesystem/introduction.html)
```
Now that we're down to tiny nits, I'd link everywhere to https:// URLs

steveloughran replied:
good catch. Done throughout the file except in the apache license header. Rendered HTML and clicked through the links.

srowen commented on hadoop-cloud/pom.xml:
```diff
@@ -0,0 +1,185 @@
+
+  <artifactId>spark-hadoop-cloud_2.11</artifactId>
+  <packaging>jar</packaging>
+  <name>Spark Project Cloud Integration</name>
```
Maybe Spark Project Cloud Object Store Integration for Hadoop? I feel like getting "Hadoop" in there is useful to somehow emphasize this is adding connectivity from HDFS APIs to object stores in particular

steveloughran replied:
As it works just as well standalone & on mesos, I'll try " through Hadoop Libraries" & see how that reads.

srowen reviewed (changes requested):
Just some minor changes now, this looks fine

steveloughran committed:
SPARK-7481: applied proofreading, moved links to https; also cut a couple of superflous blank lines

Change-Id: Iee9f0e0527de7bb875d1c2a805a0847702bb4e11

SparkQA commented:
**[Test build #76492 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76492/testReport)** for PR 17834 at commit [`32ebc8c`](https://github.com/apache/spark/commit/32ebc8cd15cd3705279fbeee1b9527abd903023d).

SparkQA commented:
**[Test build #76492 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76492/testReport)** for PR 17834 at commit [`32ebc8c`](https://github.com/apache/spark/commit/32ebc8cd15cd3705279fbeee1b9527abd903023d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76492/
Test PASSed.

srowen commented:
Merged to master
